### PR TITLE
[Movement] Collision: Potential Field

### PR DESCRIPTION
**Pull request by maltewae**
Adds the implementation for the potential field algorithm.

The PotentialField class is created for each calibration that is to be moved to a target. The other calibrations are added as obstacles.

### CODE REVIEW

File: LabExT/Movement/PathPlanning.py [hunk 1]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
...
+class PotentialField:
+    """
+    Waypoint calculation with potential field algorithm.
+    A potential field is created for each stage, with all other stages being obstacles.
+
+    This algorithm operates in 2D (x,y), assuming that the stages are on a plane.
+    """
+
+    REPULSIVE_GAIN = 100.0
+    ATTRACTIVE_GAIN = 5.0
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:143]: Probably needs to be adaptable eventually. In my simulations, I found that Attractive gain should be around 1, repulsive gain around 10000.

File: LabExT/Movement/PathPlanning.py [hunk 2]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
...
+            self.grid_size,
+            self.grid_size)
+        self.y_coords = np.arange(
+            self.grid_outline[1][0],
+            self.grid_outline[1][1] +
+            self.grid_size,
+            self.grid_size)
+        self.cx, self.cy = np.meshgrid(self.x_coords, self.y_coords)
+
+        self.attractive_potenial_field = self.ATTRACTIVE_GAIN * \
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:194]: typo

File: LabExT/Movement/PathPlanning.py [hunk 3]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
...
+        for calibration in calibrations:
+            with calibration.perform_in_chip_coordinates():
+                stage_mask = calibration.stage_polygon.stage_in_meshgrid(
+                    calibration.get_position(),
+                    self.cx,
+                    self.cy)
+
+            for ox, oy in zip(self.cx[stage_mask], self.cy[stage_mask]):
+                o_dist = np.hypot(self.cx - ox, self.cy - oy)
+                field_mask = o_dist < 5 * self.FIBER_RADIUS
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:260]: the 5 should be made a parameter

File: LabExT/Movement/PathPlanning.py [hunk 4]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
...
+                rep_field = self.REPULSIVE_GAIN * \
+                    (1.0 / o_dist - 1.0 / self.FIBER_RADIUS) ** 2
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:262]: I get smoother results in my simulation if the second term in the parentheses is omitted (i.e. assume FIBER_RADIUS -> infty for this line only)

File: LabExT/Movement/PathPlanning.py [hunk 5]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
             y_max += self.FIBER_LENGTH
 
         return x_min, x_max, y_min, y_max
+
+
+class PotentialField:
```
Comment (marcoep) [inline LabExT/Movement/PathPlanning.py:134]: Generally: this class needs unittesting covering the algorithm math behind it. E.g. generate random waypoint trajectories and check for every time that both fibers are a minimum spacing apart.

File: LabExT/Movement/PathPlanning.py [hunk 6]
```diff
@@ -126,3 +129,158 @@ def _create_outline(
...
+    """
+    Waypoint calculation with potential field algorithm.
+    A potential field is created for each stage, with all other stages being obstacles.
+
+    This algorithm operates in 2D (x,y), assuming that the stages are on a plane.
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py:167]: Changed

File: LabExT/Movement/PathPlanning.py (hunk 1 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: These are only parameters of the constructor

File: LabExT/Movement/PathPlanning.py (hunk 2 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: Fixed!

File: LabExT/Movement/PathPlanning.py (hunk 3 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: This is now parameter of the method.

File: LabExT/Movement/PathPlanning.py (hunk 4 above)
Comment (maltewae) [reply LabExT/Movement/PathPlanning.py]: I removed the term.

File: LabExT/Movement/PathPlanning.py (hunk 5 above)
Comment (marcoep) [reply LabExT/Movement/PathPlanning.py:162]: Will be solved in extra PR with simulation toolbox.

**Review by marcoep — APPROVED** [review]

**Pull request merged**